File: src/main/java/seremis/geninfusion/gui/util/GuiTab.java
package seremis.geninfusion.gui.util;

import java.util.Arrays;

import seremis.geninfusion.gui.GIGui;

public abstract class GuiTab extends GuiRectangle {

    protected int id;

    protected boolean visible = true;

    protected String name;

    public GuiTab(int id, int x, int y, int w, int h, int u, int v, String name) {
        super(x, y, w, h, u, v);
        this.name = name;
        this.id = id;
    }

    public int getId() {
        return id;
    }

    public GuiTab setVisible(boolean visible) {
        this.visible = visible;
        return this;
    }

    public boolean visible() {
        return visible;
    }

    @Override
    public void draw(GIGui gui) {
        if(visible) {
            gui.drawTexturedModalRect(gui.getLeft() + x, gui.getTop() + y, u, v, w, h);
        }
    }

    @Override
    public void drawString(GIGui gui, int mouseX, int mouseY, String str) {
        if(visible && inRect(gui, mouseX, mouseY)) {
            gui.drawHoveringString(Arrays.asList(str.split("\n")), mouseX - gui.getLeft(), mouseY - gui.getTop());
        }
    }

    public String getName() {
        return name;
    }

    @Override
    public String toString() {
        return "GuiTab[id: " + id + " x: " + x + ", y: " + y + ", width: " + w + ", heigth: " + h + " u:" + u + " v: " + v + ", name: '" + name + "']";
    }

    public abstract void drawBackground(GIGui gui, int x, int y);

    public abstract void drawForeground(GIGui gui, int x, int y);

    public void mouseClick(GIGui gui, int x, int y, int button) {}

    public void mouseMoveClick(GIGui gui, int x, int y, int button, long timeSinceClicked) {}

    public void mouseReleased(GIGui gui, int x, int y, int button) {}

    public void initGui(GIGui gui) {}

    public void keyTyped(GIGui gui, char keyTyped, int keyCode) {}
}

File: src/main/scala/seremis/geninfusion/gui/util/GuiTab.scala
package seremis.geninfusion.gui.util

import seremis.geninfusion.gui.GIGui

abstract class GuiTab(id: Int, x: Int, y: Int, w: Int, h: Int, u: Int, v: Int, name: String) extends GuiRectangle(x, y, w, h, u, v) {

    var visible: Boolean = true

    def getId: Int = id

    override def draw(gui: GIGui) {
        if (visible) {
            gui.drawTexturedModalRect(gui.getLeft + x, gui.getTop + y, u, v, w, h)
        }
    }

    override def drawString(gui: GIGui, mouseX: Int, mouseY: Int, str: String) {
        if (visible && inRect(gui, mouseX, mouseY)) {
            gui.drawHoveringString(List(str.split("\n"):_*), mouseX - gui.getLeft, mouseY - gui.getTop)
        }
    }

    def getName: String = name

    def drawBackground(gui: GIGui, x: Int, y: Int)

    def drawForeground(gui: GIGui, x: Int, y: Int)

    def mouseClick(gui: GIGui, x: Int, y: Int, button: Int) {}

    def mouseMoveClick(gui: GIGui, x: Int, y: Int, button: Int, timeSinceClicked: Long) {}

    def mouseReleased(gui: GIGui, x: Int, y: Int, button: Int) {}

    def initGui(gui: GIGui) {}

    def keyTyped(gui: GIGui, keyTyped: Char, keyCode: Int) {}

    override def toString: String = "GuiTab[id: " + id + " x: " + x + ", y: " + y + ", width: " + w + ", height: " + h + " u:" + u + " v: " + v + ", name: '" + name + "']"
}